Technote theme

jonathansick opened the pull request:
This technote theme provides the ability for Markdown, rST, or ipynb-based technotes to augment the existing LaTeX/Lander based document types for SPHEREx.

These technotes are built on the [Technote](https://technote.lsst.io) package, developed at Rubin Observatory.

Progress:

- [x] Basic theme configuration
- [x] Integrate Bibtex biliographies and source spherex.bib from bibtex
- [x] Custom sidebar to include a link to the version dashboards
- [x] Demonstration technotes in rst, md, and ipynb formats

jonathansick committed:
Update pre-commit hooks

Includes some formatting fixes because of the new Black formatter
version.

jonathansick committed:
Create spherexsphinx.conf.technote

This preset configuration provides support for SPHEREx technotes using
the technote framework (https://technote.lsst.io). The additionl
dependencies for this documentation mode are included through the
[technote] dependency extra to avoid adding unnecessary dependencies to
software projects using spherexsphinx for user guides.

The demo/md-technote/ sample project illustrates the features of SPHEREx
Sphinx technotes. It can be built by running `tox run -e demo`.

jonathansick committed:
Add common design extensions to technotes

- sphinxcontrib.mermaid provides many diagram types through Mermaid.js
- sphinx_design provides features like grids, tab groups, badges, and so
  on.
- sphinx_prompt provides a nice solution for adding command line
  examples to documentation (this wasn't previously configured, so I've
  added it as a new dependency).

jonathansick committed:
Support bibtex bibliographies in technotes

- Add sphinxcontrib.bibtex to provide support for cite roles and
  bibliography directives in technotes.
- Add Rubin's Documenteer package as an extension to provide the
  githubbibcache extension that can download a bibtex file from GitHub
- Configure SPHEREx technotes to always use the spherex.bib file from
  GitHub, and also load any .bib bibliographies from the local
  repository.

jonathansick committed:
Python 3.11 and later is required now

This requirement is put in place by Documenteer, which we use for
sourcing spherex.bib from GitHub.

gpdf commented on src/spherexsphinx/conf/_utils.py:
```diff
@@ -221,9 +221,9 @@ def set_edit_on_github(
-        html_context[
-            "github_version"
-        ] = self.config.project.github_default_branch
+        html_context["github_version"] = (
```
What's the purpose of the parentheses?

jonathansick replied:
I think a change in Black caused the formatting to change.

gpdf replied:
But it's not trying to create a tuple, right?  (In which case an extra comma would have been needed, I believe.)  It's just parenthesis-for-grouping?

jonathansick replied:
Correct; just parentheses for multi-line statement formatting, not for creating a tuple.

gpdf commented:
Looks reasonable so far; looking forward to the rest.

jonathansick committed:
Add LTD editions link to the sidebar

In the base configuration we're computing an editions_url variable that
points to the version dashboard for the document on
spherex-docs.ipac.caltech.edu.

We add this link to the sidebar by overriding the sidebar-source.html
component template that's originally provided by the technote package.

jonathansick committed:
Expand the Markdown demo with typography features

This demo is similar to the one included in the technote package with
the added SPHEREx-specific reference feature.

jonathansick committed:
Add demo technote in rst format

jonathansick committed:
Add custom spherex-technote.css for logo size

This custom CSS constrains the SPHEREx logo size and also shows the
framework for making other style adjustments in the future.

jonathansick committed:
Add support for ipynb-formatted technotes

Currently these notebooks are set to automatically execute during the
technote build if outputs are missing (or using the committed notebook if
already executed). This can be changed through myst-nb settings in
conf.py, see
https://myst-nb.readthedocs.io/en/v0.13.2/use/execute.html#triggering-notebook-execution

jonathansick committed:
Constrain to Sphinx < 9

This constraint protects against deprecated features currently used by
the pydata-sphinx-theme and technote themes.

jonathansick committed:
Run technote demos in CI

Run the demo tox environment in CI and upload the built sites as job
artifacts for review.

jonathansick commented:
The basic functionality for SPHEREx technotes on the theme-side is ready for review. In the `demo` directory you can see examples of technote projects and in the GitHub Actions runs you can find ZIP archives of the built demo sites. ([example](https://github.com/SPHEREx/spherex-sphinx/actions/runs/13039306781))

gpdf commented:
> The basic functionality for SPHEREx technotes on the theme-side is ready for review.

Just to be clear: is this going to work for all the existing document prefixes?  I.e., is "LaTeX vs tech note" orthogonal to the -MS/-OP/-IF/etc. axis?

gpdf commented on demo/ipynb-technote/technote.toml:
```diff
@@ -0,0 +1,14 @@
+[[technote.authors]]
+name.given = "Jonathan"
+name.family = "Sick"
+orcid = "https://orcid.org/0000-0003-3001-676X"
```
Is the `orcid` optional?

jonathansick replied:
Yes, most of the metadata is optional. The full schema for the `technote.toml` file is at https://technote.lsst.io/user-guide/technote-toml.html

gpdf reviewed (approved):
It all looks appropriate.

A couple of questions:

- Just to repeat: is this option (technote format, and source as MD/reST/.ipynb) available for all document categories?
- Are we going to be able to supply the recognized SPHEREx metadata that the current LaTeX-processing pipeline scans for?

jonathansick commented:
> Just to repeat: is this option (technote format, and source as MD/reST/.ipynb) available for all document categories?

Yes, you'll be able to set the `series` and `ID` other any document category. However, some of the customizations added to the LaTeX style file won't be built into this theme. So things like the project vs SSDC author and the different verification IDs don't have analogous commands here.

> Are we going to be able to supply the recognized SPHEREx metadata that the current LaTeX-processing pipeline scans for?

I'm going to have to update the documentation portal to look at the [metadata produced by Technote](https://technote.lsst.io/user-guide/metadata.html), which is distinct from the metadata produced by Lander at this point.
